cdc/sink/manager.go

```diff
@@ -0,0 +1,267 @@
+		emittedTs: checkpointTs,
+	}
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	m.tableSinks[tableID] = sink
+	return sink
+}
+
+// Close closes the Sink manager and backend Sink
+func (m *Manager) Close() error {
+	return m.backendSink.Close()
+}
+
+func (m *Manager) getMinEmittedTs() model.Ts {
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	if len(m.tableSinks) == 0 {
+		return m.getCheckpointTs()
+	}
+	minTs := model.Ts(math.MaxUint64)
+	for _, tableSink := range m.tableSinks {
+		emittedTs := tableSink.getEmittedTs()
+		if minTs > emittedTs {
+			minTs = emittedTs
+		}
+	}
+	return minTs
+}
+
+func (m *Manager) flushBackendSink(ctx context.Context) (model.Ts, error) {
+	minEmittedTs := m.getMinEmittedTs()
+	checkpointTs, err := m.backendSink.FlushRowChangedEvents(ctx, minEmittedTs)
+	if err != nil {
+		return m.getCheckpointTs(), errors.Trace(err)
+	}
+	atomic.StoreUint64(&m.checkpointTs, checkpointTs)
+	return checkpointTs, nil
+}
+
+func (m *Manager) destroyTableSink(tableID model.TableID) {
+	m.tableSinksMu.Lock()
+	defer m.tableSinksMu.Unlock()
+	delete(m.tableSinks, tableID)
+}
+
+func (m *Manager) getCheckpointTs() uint64 {
+	return atomic.LoadUint64(&m.checkpointTs)
+}
+
+type tableSink struct {
+	tableID model.TableID
+	manager *Manager
+	buffer  []*model.RowChangedEvent
+	// emittedTs means all of events which of commitTs less than or equal to emittedTs is sent to backendSink
+	emittedTs model.Ts
+}
+
+func (t *tableSink) Initialize(ctx context.Context, tableInfo []*model.SimpleTableInfo) error {
+	// do nothing
+	return nil
+}
+
+func (t *tableSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	t.buffer = append(t.buffer, rows...)
+	return nil
+}
+
+func (t *tableSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	// the table sink doesn't receive the DDL event
+	return nil
+}
+
+func (t *tableSink) FlushRowChangedEvents(ctx context.Context, resolvedTs uint64) (uint64, error) {
+	i := sort.Search(len(t.buffer), func(i int) bool {
+		return t.buffer[i].CommitTs > resolvedTs
+	})
+	if i == 0 {
+		atomic.StoreUint64(&t.emittedTs, resolvedTs)
+		return t.manager.flushBackendSink(ctx)
+	}
+	resolvedRows := t.buffer[:i]
+	t.buffer = t.buffer[i:]
+
+	err := t.manager.backendSink.EmitRowChangedEvents(ctx, resolvedRows...)
+	if err != nil {
+		return t.manager.getCheckpointTs(), errors.Trace(err)
+	}
+	atomic.StoreUint64(&t.emittedTs, resolvedTs)
+	return t.manager.flushBackendSink(ctx)
+}
+
+func (t *tableSink) getEmittedTs() uint64 {
+	return atomic.LoadUint64(&t.emittedTs)
+}
+
+func (t *tableSink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
+	// the table sink doesn't receive the checkpoint event
+	return nil
+}
+
+func (t *tableSink) Close() error {
+	t.manager.destroyTableSink(t.tableID)
+	return nil
+}
+
+type bufferSink struct {
+	Sink
+	buffer chan struct {
+		rows       []*model.RowChangedEvent
+		resolved   bool
```

Comment:
We can remove the `resolved` field, and check whether it is a resolved event by `rows == nil`